bee-message/bee-byte-cost-derive/src/enum_info.rs

```diff
@@ -0,0 +1,69 @@
+// Copyright 2021-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    builder::ByteCostBuilder,
+    field_info::{fields_to_info, FieldInfo},
+    weight::filter_attrs,
+};
+
+use proc_macro2::TokenStream;
+use quote::{format_ident, quote, ToTokens};
+use syn::{Ident, Result, Variant};
+
+pub(crate) struct VariantInfo {
+    ident: Ident,
+    fields: Vec<FieldInfo>,
+}
+
+impl VariantInfo {
+    pub(crate) fn new(variant: &Variant) -> Result<Self> {
+        if filter_attrs(&variant.attrs).next().is_some() {
+            return Err(syn::Error::new(
+                variant.ident.span(),
+                "weight types cannot be applied to variants of an enum",
+            ));
+        }
+
+        Ok(Self {
+            fields: fields_to_info(&variant.fields)?,
+            ident: variant.ident.clone(),
+        })
+    }
+}
+
+impl ToTokens for VariantInfo {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let ident = &self.ident;
+        let variant_fields = (0..self.fields.len()).map(|n| format_ident!("field_{}", n));
+        let variant_fields_2 = variant_fields.clone();
```

Comment:
would this mean that `|n| format_ident!("field_{}", n)` is invoked twice? in that sense it might be faster to collect `variant_fields` and then `let variant_fields_2 = &variant_fields;` or something.